### ISSUE TITLE
Code inside `eval` triggers deprecation, results in `ExcludeList::isExcluded(): Argument #1 ($file) must be of type string, null given`

### ISSUE THREAD

**Issue by ruudk** [sebastianbergmann/phpunit#6043]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 11.5-dev
| PHP version         | 8.4
| Installation Method | Composer

#### Summary

When `eval`'d code triggers a deprecation, the following error happens:
> PHPUnit\Util\ExcludeList::isExcluded(): Argument #1 ($file) must be of type string, null given, called in /Volumes/CS/opensource/Twig/vendor/phpunit/phpunit/src/Runner/ErrorHandler.php on line 403")

#### Current behavior

This was introduced in https://github.com/sebastianbergmann/phpunit/commit/5d6dc50e2dab3834c40a209d06a89d6d1f18d61e#diff-c59b939cd44edcda5d8a71c38f90461826dc2abbb5267f343fd7db0fa8fe1784

https://github.com/sebastianbergmann/phpunit/blob/dac4a5f6e376f801617128668f8cd828a59336c5/src/Runner/ErrorHandler.php#L403-L405

Here, it should handle the fact that `file` cannot be there. 

#### How to reproduce

PR where the issue is visible: https://github.com/twigphp/Twig/pull/4472

When running the debugger, you can see that `file` is not part of the stack trace frame when the code is eval'd.
 
![Screenshot 2024-11-25 at 09 52 26@2x](https://github.com/user-attachments/assets/97ed14fc-5d3b-4d6f-8fb8-dd0b4a105639)

#### Expected behavior

No error.

**Comment by sebastianbergmann** [comment]
Thank you for your report.

Please provide a **minimal**, **self-contained**, reproducing test case that shows the problem you are reporting.

Without such a minimal, self-contained, reproducing test case I will not be able to investigate this issue.

**Comment by ruudk** [comment]
I am able to create a test that triggers a deprecation. But I cannot mimic the scenario that happens with Twig. Twig uses `yield` and generators on an eval'd template. 

But even with this, I always receive the `file`:
```php
final class Test extends TestCase
{
    public static function testEval(): void
    {
        $code = <<<'PHP'
function main() {
    yield fn() => \PHPUnit\TestFixture\Bug6043\trigger_deprecation('deprecation triggered from within eval', E_USER_DEPRECATED);
}
foreach (main() as $item) {
    echo $item();
}
PHP;

        eval($code);
    }
}
```

**Comment by sebastianbergmann** [comment]
The example given in https://github.com/sebastianbergmann/phpunit/issues/6043#issuecomment-2497407111 is not self-contained, as it uses a function named `\PHPUnit\TestFixture\Bug6043\trigger_deprecation` which is not provided.

I have changed the example to make it run like so:
```php
<?php declare(strict_types=1);
use PHPUnit\Framework\TestCase;

final class Test extends TestCase
{
    public function testEval(): void
    {
        $code = <<<'PHP'
function main() {
    yield fn() => trigger_error('message', E_USER_DEPRECATED);
}

foreach (main() as $item) {
    $item();
}
PHP;

        eval($code);

        $this->assertTrue(true);
    }
}
```
With the above, I get
```
❯ ./phpunit Test.php --display-deprecations
PHPUnit 11.5-g7e4dffb7d7 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.4.1
Configuration: /usr/local/src/phpunit/phpunit.xml

D                                                                   1 / 1 (100%)

Time: 00:00.028, Memory: 12.00 MB

1 test triggered 1 deprecation:

1) /usr/local/src/phpunit/Test.php(18) : eval()'d code:2
message

Triggered by:

* Test::testEval
  /usr/local/src/phpunit/Test.php:6

OK, but there were issues!
Tests: 1, Assertions: 1, Deprecations: 1.
```
This looks correct to me, but you mentioned that "even with this, [you] always receive the `file`", so that is probably expected and what you meant.

I implemented a safe guard to not use the `file` index of a stack frame array when it is not available in 0009ef0f7f8f6fd0d7abfce7f1d0520de2868e09.

Please let me know whether this solves the issue for you. Thanks!

**Comment by ruudk** [comment]
I just ran the Twig suite on latest phpunit main again and can confirm the problem is gone now. 

Thanks for the fast fix and pointing out the problem that I was using the dev dependency without knowing it.
